docs/contributing/writing-instrumentation-module.md

```diff
@@ -360,3 +360,129 @@ For example:
 ```
 
 [suppress]: https://opentelemetry.io/docs/instrumentation/java/automatic/agent-config/#suppressing-specific-auto-instrumentation
+
+## Use non-inlined advice code with `invokedynamic`
+
+Using non-inlined advice code is possible thanks to the `invokedynamic` instruction, this strategy
+is referred as "indy" in reference to this, by extension "indy modules" are the instrumentation
+modules using this instrumentation strategy.
+
+The most common way to instrument code with bytebuddy relies on inlining, this strategy will be
+referred as "inlined" strategy in opposition to "indy".
+
+### Indy modules and transition
+
+Having all instrumentation rely on native "indy" instrumentation is a tedious task and can't be
+achieved in a "big bang" step, we thus have to use intermediate steps.
+
+Instrumentation modules that are "indy native" must have their `InstrumentationModule#isIndyModule`
+implementation return `true`. Also, all the instrumentation advice methods annotated with
+`@Advice.OnMethodEnter` or `@Advice.OnMethodExit` must have the `inlined = false` explicitly set.
+
+The `otel.javaagent.experimental.indy` (default `false`) configuration option allows to opt-in for
+using "indy". When set to `true`, the `io.opentelemetry.javaagent.tooling.instrumentation.indy.AdviceTransformer`
+will transform advices automatically to make them "indy native".
+
+This configuration is automatically enabled in CI when `test indy` label is added to a pull-request
+or when the `-PtestIndy=true` parameter is added to gradle.
+
+In order to preserve compatibility with both instrumentation strategies, we have to omit the `inlined = false`
+from the advice method annotations.
+
+We have two sets of instrumentation modules:
+- "indy compatible": compatible with both "indy" and "inlined", do not override `isIndyModule`
+- "inlined only": only compatible with "inlined", `isIndyModule` returns `false`.
+
+The first step of the migration is to move all the "inlined only" to the "indy compatible" category
+by refactoring them with the limitations described below.
+
+Once everything is "indy compatible", we can evaluate changing the default value of `otel.javaagent.experimental.indy`
+to `true` and make it non-experimental.
```

Comment:
I'd assume that the config option remains experimental until it eventually is removed?